### PR TITLE
feat(GraphQL): Graphql Custom API

### PR DESCRIPTION
**Pull request by abias1122**
## What does this do?
Creates a custom GraphQL API to hopefully provide a better experience than the auto generated one.

## How was it tested?
I ran the code locally and queried the API to make sure I got the expected results and that nothing broke.

## Is there a Github issue this is resolving?
[Yes](https://github.com/5e-bits/5e-srd-api/issues/238)

## Was any impacted documentation updated to reflect this change?
I will update the documentation in a later commit. Just getting the PR up now

## Here's a fun image for your troubles
random photo - update me

### CODE REVIEW

File: src/graphql/resolvers/proficiencyResolver.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    if (url.includes('ability-scores')) return await AbilityScoreModel.findOne({ url }).lean();
+    if (url.includes('equipment-categories'))
+      return await EquipmentCategoryModel.findOne({ url }).lean();
+    if (url.includes('equipment')) return await EquipmentModel.findOne({ url }).lean();
+    if (url.includes('skills')) return await SkillModel.findOne({ url }).lean();
+  },
+  type: proficiency =>
+    proficiency.type
+      .toUpperCase()
+      .replace("'", '')
```
[the remark this answers is not in the document]
Comment (abias1122) [reply src/graphql/resolvers/proficiencyResolver.js:36]: I believe I ran into problems using `replaceAll` and it ended up being unnecessary anyway.  That and the proficiency type is just what's defined in the db's json file, so it's not like we're validating input or anything.

[the remark this answers is not in the document]
Comment (bagelbits) [reply src/graphql/resolvers/proficiencyResolver.js:36]: You could also just do `.replace(/'/g, '')`

[the remark this answers is not in the document]
Comment (bagelbits) [reply src/graphql/resolvers/proficiencyResolver.js:36]: You do the same thing in a few other places. May as well copy it here.

**Review by bagelbits — COMMENTED** [review]
I have a few questions before I approve it, but I think it looks pretty good.

It's a shame we can't auto generate this anymore. But I guess that was inevitable.

File: src/graphql/typeDefs.graphql [hunk 2]
```diff
@@ -0,0 +1,1439 @@
+"""
```
[the remark this answers is not in the document]
Comment (abias1122) [reply src/graphql/typeDefs.graphql:1]: I'm honestly not sure.

File: src/graphql/resolvers/classResolver.js [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+        ...option.choice,
+        from: {
+          ...option.choice.from,
+          equipment_category: await EquipmentCategoryModel.findOne({
+            index: option.choice.from.equipment_category.index,
+          }).lean(),
+        },
+      },
+    };
+  }
```
[the remark this answers is not in the document]
Comment (abias1122) [reply src/graphql/resolvers/classResolver.js:30]: I could see if I can extract a function for it.

File: src/graphql/resolvers/common.js [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+  if (equipment.armor_class) return 'Armor';
+  if (equipment.weapon_category) return 'Weapon';
+  if (equipment.vehicle_category) return 'Vehicle';
+  return null;
+};
+
+const resolveSpellOrderBy = value =>
+  value === 'AREA_OF_EFFECT_SIZE' ? 'area_of_effect.size' : value.toLowerCase();
+
+export const resolveSpells = async (args, baseFilters) => {
```
[the remark this answers is not in the document]
Comment (abias1122) [reply src/graphql/resolvers/common.js:40]: Yes.

File: src/graphql/resolvers/ammunitionResolver.js [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import { gearFieldResolvers } from './common';
```
[the remark this answers is not in the document]
Comment (abias1122) [reply src/graphql/resolvers/ammunitionResolver.js]: Whichever you prefer. I have no preference regarding this.

File: src/graphql/resolvers/queryResolver.js [hunk 6]
```diff
@@ -0,0 +1,646 @@
+import {
```
[the remark this answers is not in the document]
Comment (abias1122) [reply src/graphql/resolvers/queryResolver.js:1]: When you create a GraphQL schema, you have to have a `resolvers` object that can resolve all the types in your typedefs. This resolver happens to be for Query, which is the top level thing that gets resolved.

File: src/graphql/resolvers/scalars/monsterSubtypeFilterResolver.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+  'ORC',
+  'SAHUAGIN',
+  'TITAN',
+  'KOBOLD',
+  'GNOLL',
+  'GRIMLOCK',
+  'LIZARDFOLK',
+  'GNOME',
+];
+
```
[the remark this answers is not in the document]
Comment (abias1122) [reply src/graphql/resolvers/scalars/monsterSubtypeFilterResolver.js:22]: I can sort them, however these enum filters are only used for input so I don't think we'd really get anything from doing that.

File: src/graphql/resolvers/scalars/sizeFilterResolver.js [hunk 8]
```diff
@@ -0,0 +1,9 @@
+import { createEnumScalarType } from './enumFilterResolver';
+
+const sizes = ['MEDIUM', 'LARGE', 'TINY', 'HUGE', 'SMALL', 'GARGANTUAN'];
```
[the remark this answers is not in the document]
Comment (abias1122) [reply src/graphql/resolvers/scalars/sizeFilterResolver.js:3]: I can change the order.

File: src/graphql/resolvers/iEquipmentBaseResolver.js [hunk 9]
```diff
@@ -0,0 +1,11 @@
+import { resolveEquipmentType } from './common';
+
+const IEquipmentBase = {
```
[the remark this answers is not in the document]
Comment (abias1122) [reply src/graphql/resolvers/iEquipmentBaseResolver.js:3]: `IEquipmentBase` can be an piece of equipment or a magic item.

`IEquipment` is a piece of equipment. There are various types that inherit `IEquipment` like weapon and armor. They still all use the equipment mongo collection though.

**Review by bagelbits — APPROVED** [review]

**Pull request merged**